No string term2

Kakadu opened the pull request.

georgiy-belyanin committed:
Eliminate string terms

georgiy-belyanin committed:
solver: evaluate regex with constants on presimpl

This patch makes the expression simplifier more advanced by allowing it
to evaluate expressions in the form of `(str.in.re <str const> <re>)`.

It is especially important in the cases when the <str const> is obtained
from earlier simplifications allowing not to run the NFA solver
completely. A little bit more complex example where it works so is as
follows.

```
(set-logic QF_S)

(declare-const x String)

(assert (= x "abc"))
(assert (str.in_re x (str.to_re "abc")))

(check-sat)
```

Kakadu committed:
Fix strings after cherry-pick

Signed-off-by: Kakadu <Kakadu@pm.me>

Kakadu committed:
Promote some tests

Signed-off-by: Kakadu <Kakadu@pm.me>

Kakadu committed:
Repair 2 more tests (forgotten -flat switch)

Signed-off-by: Kakadu <Kakadu@pm.me>

Kakadu committed:
Simplify 'var = int_of_string ...' properly

Signed-off-by: Kakadu <Kakadu@pm.me>

Kakadu committed:
Minor polishing. Fixup?

Signed-off-by: Kakadu <Kakadu@pm.me>

Kakadu committed:
Promote some tests

Signed-off-by: Kakadu <Kakadu@pm.me>

Kakadu committed:
Promote tests

Signed-off-by: Kakadu <Kakadu@pm.me>

Kakadu committed:
WTF test 10001 works 30 seconds

Signed-off-by: Kakadu <Kakadu@pm.me>

georgiy-belyanin reviewed (dismissed):
Thank you for working on this!

georgiy-belyanin committed:
solver: support

The patch fixes substitutions of the complex terms inside the regular
string constraints. Previously, only atoms and constant might be
substituted. Now arbitrary integer equations are supported to be
introduced by the simplifications.

For instance, see the following example that previously had failed with
an error.

```
(set-logic QF_S)
(declare-fun x () String)
(assert (str.in_re x (re.++ (str.to_re "12345") (re.* (re.range "0" "9")) )))
(assert (= (mod (mod (str.to_int x) 331) 179) 0))
(check-sat)
```

georgiy-belyanin committed:
Try using GADTs, wip

mklstarch commented:
Now part of #192
